---
title: "QuaternaryProd"
author: "Carl Tony Fakhry, Ping Chen and Kourosh Zarringhalam"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
        

vignette: >
  %\VignetteIndexEntry{<span style="color:red">QuaternaryProdVignette</span>}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

\(\indent\) \(\indent\) \(\indent\) \(\indent\) A signed causal graph is a directed graph where the edges are signed and the signs indicate the direction of effect of the source node on the target node (the signs are either + or -). **QuaternaryProd** is a package for computing the Quaternary Dot Product Scoring Statistic (or simply the Quaternary Statistic) for signed causal graphs. The Quaternary Dot Product Scoring Statistic is a generalization of the Ternary Dot Product Scoring Statistic [1] which allows for ambiguities to arise in a signed causal graph. Ambiguities arise when a source node can affect a target node in two different ways or if the direction of causality is unknown. We will first provide some background, and then we will apply the statistic to STRINGdb which is a publicly available biological network.

## Introduction
\(\indent\) \(\indent\) \(\indent\) \(\indent\)  The Quaternary Dot Product Scoring Statistic [2] is a goodness of fit test for examining how well the predictions of a signed and directed causal graph predict on newly realized experimental data. Given a source node \(s\) in a signed causal graph, let \(q_p\), \(q_m\) and \(q_r\) denote the number of target nodes which are increased, decreased and regulated by the source node respectively. Similarly, let \(q_z\) denote the set of target nodes in the causal network which do not share a relation with \(s\) i.e which are not affected by \(s\). Regulated relations occur when a source node regulates a target node without knowing the direction of causality or if an ambiguity in direction of causality occurs. An ambiguity can occur if a source node, according to a given network, shares both increase and decrease relations with the same target node. 
Next, Suppose we run some experiments on entities which are target nodes in the network. Let \(n_p\), \(n_m\) and \(n_z\) denote the set of values which are increased, decreased and remain unchanged in the experimental values respectively. For the source node \(s\), we can tabulate the predictions from the network vs. the experimental values:

\begin{table}[!htbp]
\centering
\begin{tabular}{l l l l l}
\hline
 & Observed \(+\)  & Observed \(-\) & Observed \(0\) &  Total \\ \hline
Predicted \(+\) & \(n_{pp}\) & \(n_{pm}\)  & \(n_{pz}\)  & \(q_p\) \\
Predicted \(-\) & \(n_{mp}\)  & \(n_{mm}\)  & \(n_{mz}\)  & \(q_m\) \\
Predicted \(r\) & \(n_{rp}\)  & \(n_{rm}\)  & \(n_{rz}\)  & \(q_r\) \\
Predicted \(0\) & \(n_{zp}\) & \(n_{zm}\) & \(n_{zz}\) & \(q_z\) \\
Total & \(n_p\)  & \(n_{m}\)   & \(n_z\) & \(T\) \\
\hline
\end{tabular}
\caption{Tabulation of predictions from network edges vs. observations from experimental results.}
\end{table}

\(n_{pp}\) denotes the number of target nodes which \(s\) is predicted to increase by the network and were indeed increased in  experimental values; \(n_{pm}\) the number of target nodes which \(s\) is predicted to increase and were decreased in experimental values; \(n_{pz}\) is the number of target nodes which \(s\) is predicted to increase and were unchanged in experimental values. Similar interpretation follows for all other entries of the table. 
The probability of a table follows the Quaternary Dot Product distribution which is given by:
\begin{align}
P(\text{Table}) &= \frac{ \binom{q_p}{n_{pp},n_{pm},n_{pz}} \binom{q_m}{n_{mp},n_{mm},n_{mz}}\binom{q_z}{n_{zp},n_{zm},n_{zz}}\binom{q_r}{n_{rp},n_{rm},n_{rz}}}{\binom{T}{n_p,n_m,n_z}}.
\end{align}
Note, since the predictions by the network and the experimental values are fixed, then the table has 6 degrees of freedom \(n_{pp}\), \(n_{mm}\), \(n_{rp}\), \(n_{rm}\), \(n_{mp}\) and \(n_{pm}\). The score \(S\) to measure the goodness of fit is given by:
\begin{align}
S(\text{Table}) &= n_{pp} + n_{mm} + n_{rp} + n_{rm} - (n_{mp} + n_{pm}) 
\end{align}
which is the sum of the good predictions (i.e \(n_{pp}\), \(n_{mm}\), \(n_{rp}\) and \(n_{rm}\)) minus the bad predictions (i.e \(n_{mp}\) and \(n_{pm}\)). To compute the probability of a score, we sum the probabilites of all tables with score \(S\) as follows:
\begin{align}
P(S) &= \sum_{P(\text{Table}) = S } P(\text{Table}).
\end{align}

## Functionality
\(\indent\) \(\indent\) \(\indent\) \(\indent\)   **QuaternaryProd** provides different functions for computing the probability of a score, probability mass function, p-value of a score and the domain of the Quaternary Dot Product Scoring Statistic. The probability mass function can be computed if given the margins of the table. 
  
```{R, message=FALSE }
library(QuaternaryProd)

# Compute the probability mass function
pmf <- QP_Pmf(q_p = 20, q_m = 20, q_z = 20, q_r = 0, n_p = 20, n_m = 20, n_z = 20)

# Plot the mass function
plot(names(pmf), pmf, col="blue", xlab = "scores", ylab = "probabilities")
lines(names(pmf), pmf, col = "blue")
```  
  
The package contains optimized functions for computing the p-value of a score. To compute the p-value of score we can use the following:

```{R, message=FALSE }
# Get the p-value of score 5
pval <- QP_Pvalue(score = 5, q_p = 20, q_m = 20, q_z = 20, q_r = 0, 
                                                     n_p = 20, n_m = 20, n_z = 20)
pval

# Compue the p-value only if it is statistically significant otherwise
# return -1
pval <- QP_SigPvalue(score = 5, q_p = 20, q_m = 20, q_z = 20, q_r = 0, 
                                                     n_p = 20, n_m = 20, n_z = 20)
pval
```  

If the user is only interested in obtaining statistically significant p-values, then `QP_SigPvalue` is optimized for this purpose. In either case, the user is advised to compute the p-value of a score using the previous two functions which will be faster than computing the entire probability mass function and then computing the p-value.
Finally, it is possible to also compute the probabilities of scores individually using `QP_Probability` and the support of the distribution using `QP_Support`. 
Since this package is written to the benefit of bioinformaticians, we will provide an example on how to apply this statistic to a publicly available network.
One bioinformatic application is to test how well protein-protein causal networks can predict the results in gene expression data. In the last section of this Vignette, we present an example of computing this statistic over the STRINGdb network and given gene expression data.   
  
## Functionality for working with the Homo sapien causal network from STRINGdb

\(\indent\) \(\indent\) \(\indent\) \(\indent\) Here we provide functionality for using **QuaternaryProd** with the STRINGdb Homo Sapien causal network version 10.5 provided under the creative commons license. 

### Compute Pvalues Over the Network
\indent  Given new gene expression data, we can compute the scores and p-values for all source nodes in the STRINGdb Homo Sapien causal network using the specialized `RunCRE_HSAStringDB` function. We use the gene expression datasets that were used in [1]. The datasets contain the c-Myc, H-Ras and E2F3 expression signatures.

```{r, message=FALSE}
library(QuaternaryProd)
library(fdrtool)

# Gene expression data
gene_expression_data1 <- system.file("extdata", "e2f3_sig.txt",
                                      package = "QuaternaryProd")
gene_expression_data1 <- read.table(gene_expression_data1, sep = "\t",
                                      header = TRUE, stringsAsFactors = FALSE)
gene_expression_data2 <- system.file("extdata", "myc_sig.txt"
                                      , package = "QuaternaryProd")
gene_expression_data2 <- read.table(gene_expression_data2, sep = "\t",
                                      header = TRUE, stringsAsFactors = FALSE)
gene_expression_data3 <- system.file("extdata", "ras_sig.txt"
                                      , package = "QuaternaryProd")
gene_expression_data3 <- read.table(gene_expression_data3, sep = "\t",
                                      header = TRUE, stringsAsFactors = FALSE)

# Remove duplicated entrez ids in the gene expression data and rename 
# column names appropriately
names(gene_expression_data1) <- c("entrez", "pvalue", "fc")
gene_expression_data1 <- gene_expression_data1[!duplicated(gene_expression_data1$entrez),]

names(gene_expression_data2) <- c("entrez", "pvalue", "fc")
gene_expression_data2 <- gene_expression_data2[!duplicated(gene_expression_data2$entrez),]

names(gene_expression_data3) <- c("entrez", "pvalue", "fc")
gene_expression_data3 <- gene_expression_data3[!duplicated(gene_expression_data3$entrez),]

# Compute Quaternary Dot Product Scoring Statistic for all source nodes based
# on new gene expression data. 
CRE_results <- RunCRE_HSAStringDB(gene_expression_data1, method = "Quaternary",
                                   fc.thresh = log2(1.3), pval.thresh = 0.05)
# Get FDR corrected p-values
CRE_results["qvalue"] <- p.adjust(CRE_results$pvalue, method = "fdr")
CRE_results[1:8, c("uid","symbol","regulation","pvalue","qvalue")]

# CRE_results2 <- RunCRE_HSAStringDB(gene_expression_data2, method = "Ternary",
#                                    fc.thresh = log2(1.3), pval.thresh = 0.05)
# # Get FDR corrected p-values
# CRE_results2["qvalue"] <- p.adjust(CRE_results2$pvalue, method = "fdr")
# CRE_results2[1:400, c("uid","symbol","regulation","pvalue","qvalue")]


# CRE_results3 <- RunCRE_HSAStringDB(gene_expression_data3, method = "Quaternary",
#                                    fc.thresh = log2(1.3), pval.thresh = 0.05)
# # Get FDR corrected p-values
# CRE_results3["qvalue"] <- p.adjust(CRE_results3$pvalue, method = "fdr")
# CRE_results3[1:400, c("uid","symbol","regulation","pvalue","qvalue")]

```

`RunCRE_HSAStringDB` returns a data frame containing all the source nodes of the String causal network, all of which had their respective scores and p-values computed. The source nodes are ordered in increasing order of the p-values (Note: details on the columns of the data frame returned can be found in the help page for `RunCRE_HSAStringDB`).

## References

[1] Chindelevitch et al. (2012). Assessing statistical significance in causal graphs. BMC Bioinformatics, Volume 3, Issue 1, 2012, Page 35.

[2] Carl Tony Fakhry, Parul Choudhary, Alex Gutteridge, Ben Sidders, Ping Chen, Daniel Ziemek, and Kourosh Zarringhalam. Interpreting transcriptional changes using causal graphs: new methods and their practical utility on public networks. BMC Bioinformatics, 17:318, 2016. ISSN 1471-2105. doi: 10.1186/s12859-016-1181-8.

[3] Franceschini, A (2013). STRING v9.1: protein-protein interaction networks, with increased coverage and integration. In:'Nucleic Acids Res. 2013 Jan;41(Database issue):D808-15. doi: 10.1093/nar/gks1094. Epub 2012 Nov 29'.



